### ISSUE TITLE
Mapservers WMS layer wrong placed tiles

### ISSUE THREAD

**Issue by am2222** [SharpMap/SharpMap#85]
Hi,
I am working on a wms layer. here is my code
```C#
string url = "http://gmis.jrc.ec.europa.eu/webservices/4km/wms/modisa?DATASET=chla&VERSION=1.3.0&";

if(wms==null)
    wms = new WmsLayer("Metacarta", url);

wms.AddLayer("GMIS_A_CHLA_07_2002");
wms.SetImageFormat(wms.OutputFormats[0]);
wms.ContinueOnError = false;
//Skip rendering the WMS Map if the server couldn't be requested (if set to false such an event would crash the app)
wms.TimeOut = 200000; //Set timeout to 5 seconds
wms.SRID = 4326;
mapBox1.Map.BackgroundLayer.Add( AsyncLayerProxyLayer.Create(wms));
return;
```

When I add it to map it shows tiles in wrong places

<img width="898" alt="sketch" src="https://user-images.githubusercontent.com/4707503/37877151-e417b956-306b-11e8-84b5-1b29556d80ae.png">

My maps SRID is `4326` and layers srid as you can see in its get capabilities is `4326`,as you can see if we stretch tiles it seems they place in their right place. I am still working to find possible solutions. I'd be thankful if you give me some hints about this problem. thanks

**Comment by am2222** [comment]
Well, I think there is sth wrong with maps envelop. 
I tried to get request URL from WMS layer. here is what it generates

`http://gmis.jrc.ec.europa.eu/cgi-bin/mapserv?map=/srv/www/htdocs/wms/wms_4km_modisa-chla.map&REQUEST=GetMap&BBOX=-450.330882352941,-180,450.330882352941,180&WIDTH=1361&HEIGHT=544&LAYERS=GMIS_A_CHLA_07_2002&FORMAT=image/jpeg&CRS=EPSG:4326&VERSION=1.3.0&Styles=&TRANSPARENT=True`


I noticed about `bbox` it give `45.0` to `450` which make the problem. The above url must be as

`http://gmis.jrc.ec.europa.eu/cgi-bin/mapserv?map=/srv/www/htdocs/wms/wms_4km_modisa-chla.map&REQUEST=GetMap&BBOX=-45.330882352941,-180,45.330882352941,180&WIDTH=1361&HEIGHT=544&LAYERS=GMIS_A_CHLA_07_2002&FORMAT=image/jpeg&CRS=EPSG:4326&VERSION=1.3.0&Styles=&TRANSPARENT=True`

I have set map.SRID=4326 but it calculates envlop wrong

<img width="519" alt="untitled" src="https://user-images.githubusercontent.com/4707503/37878666-c5a92a0c-3081-11e8-84ad-93dd03b1f970.png">

**Comment by am2222** [comment]
well I found that the problem is with `URL` and with `bbox` parameter. in fact instead of `-85.0404114621602,-180,85.0404114621602,180` The generated URL is `-180,-85.0404114621602,180,85.0404114621602` I guess I have to use a `Matrix` to transform those values. but  I have no idea if it works or not.

**Comment by am2222** [comment]
I managed to fix the problem using this wrapper.

```C#
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Text;
using System.Threading.Tasks;
using GeoAPI.Geometries;
using SharpMap;

namespace SSTExtracter.Img
{
    class Wmsinvers: SharpMap.Layers.WmsLayer
    {

        public Wmsinvers(string layername, string url)
            : base(layername,  url)
        {
        }


        public override string GetRequestUrl(Envelope box, Size size)
        {
            string url= base.GetRequestUrl(box, size);

            
            var right = new StringBuilder();
            right.AppendFormat(Map.NumberFormatEnUs, "REQUEST=GetMap&BBOX={0},{1},{2},{3}",
                                 box.MinY, box.MinX,  box.MaxY,box.MaxX);
            var wrong = new StringBuilder();
            wrong.AppendFormat(Map.NumberFormatEnUs, "REQUEST=GetMap&BBOX={0},{1},{2},{3}",
                                box.MinX, box.MinY, box.MaxX, box.MaxY);
            url=url.Replace(wrong.ToString(), right.ToString());
            return url;
        }
    }
```

I have no idea about other possible methods. or even why this happend. but it works now

**Comment by FObermaier** [comment]
@am2222 , you are tiling the WMS using `AsyncLayerProxyLayer`. Out of curiosity, the plain `WmsLayer` works correct? What is the extent of the `WmsLayer`?

**Comment by am2222** [comment]
@FObermaier  I tried both `AsyncLayerProxyLayer` and `wmslayer` and none of them worked. when I changed `bbox` parameters both of them started to work fine. It seems `bbox` parameters were wrong for this specific `wms` layer. I have no idea why this happened. I have tested with other wms layer and this error does not happened. the extent of wms layer as it gets from `getcapabilities` is 


```
EX_GeographicBoundingBox>
<westBoundLongitude>-180</westBoundLongitude>
<eastBoundLongitude>180</eastBoundLongitude>
<southBoundLatitude>-90</southBoundLatitude>
<northBoundLatitude>90</northBoundLatitude>
</EX_GeographicBoundingBox>
<BoundingBox CRS="EPSG:4326" minx="-90" miny="-180" maxx="90" maxy="180"/>
```

The getcapabilities url is as following

`http://gmis.jrc.ec.europa.eu/webservices/4km/wms/modisa/?DATASET=chla&VERSION=1.3.0&`

the intresting part is that qgis loads it with no errors. it seems it some how finds out that parameters of bbox must be changed.
